test_util/azure.py

```diff
@@ -0,0 +1,198 @@
+        """
+        assert master_prefix != agent_prefix, 'Master and agents must have unique prefixs'
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', master_prefix),\
+            'Invalid master DNS prefix: {}'.format(master_prefix)
+        assert re.match('^[a-z][a-z0-9-]{1,61}[a-z0-9]$', agent_prefix),\
+            'Invalid agent DNS prefix: {}'.format(agent_prefix)
+
+        deployment_name = DEPLOYMENT_NAME.format(group_name)
+
+        # Resource group must be created before validation can occur
+        if azure_wrapper.rmc.resource_groups.check_existence(group_name):
+            raise Exception("Group name already exists / taken: {}".format(group_name))
+        log.info('Starting resource group_creation')
+        azure_wrapper.rmc.resource_groups.create_or_update(
+            group_name,
+            ResourceGroup(location=location))
+
+        template_params = {
+            'sshRSAPublicKey': public_key,
+            # must be lower or validation will fail
+            'masterEndpointDNSNamePrefix': master_prefix.lower(),
+            'agentEndpointDNSNamePrefix': agent_prefix.lower(),
+            'linuxAdminUsername': admin_name,
+            'agentVMSize': vm_size,
+            'agentCount': agent_count,
+            'nameSuffix': name_suffix,
+            'oauthEnabled': repr(oauth_enabled).lower(),
+            # oauth uses string, vm diagnostic uses bool
+            'enableVMDiagnostics': vm_diagnostics_enabled}
+        log.info('Provided template parameters: {}'.format(template_params))
+        # add worthless required 'value' fields
+        template_parameters = {k: {'value': v} for k, v in template_params.items()}
+        deployment_properties = DeploymentProperties(
+            template_link=TemplateLink(uri=template_uri),
+            mode=DeploymentMode.incremental,
+            parameters=template_parameters)
+        # Use RPC against azure to validate the ARM template is well-formed
+        log.info('Validating template deployment')
+        try:
+            result = azure_wrapper.rmc.deployments.validate(
+                group_name, deployment_name, properties=deployment_properties)
+            if result.error:
+                for details in result.error.details:
+                    log.error('{}: {}'.format(details.code, details.message))
+                raise Exception("Template verification failed!")
+        except:
+            # no point in leaving a fresh resource group with nothing in it around
+            log.exception('Deployment is invalide, deleting provisioned resource group')
+            azure_wrapper.rmc.resource_groups.delete(group_name)
+            raise
+
+        log.info('Starting template deployment')
+        azure_wrapper.rmc.deployments.create_or_update(
+            group_name, deployment_name, deployment_properties)
+
+        new_deploy = cls(group_name, azure_wrapper)
+        return new_deploy
+
+    def wait_for_deployment(self, timeout=60 * 60):
```

Comment:
these timeouts are radiccally changed from the roiginal code (45 attempts, each one minute apart). Please do the old behavior. Can talk about changing it, but this PR already has a lot moving in it, and I'd rather keep things as small as possible.